### PR TITLE
Add retry_allocator for gpu

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR adds `RetryAllocator` for CUDA allocator. This is designed for fault tolerance if there are multiple threads want to allocate GPU memory in the same time.

See [CE Debug](http://ce.paddlepaddle.org:8080/viewLog.html?buildId=56975&tab=buildResultsDiv&buildTypeId=PaddleModesl_ModelDebug_Model11Build) here. Although `model_image_classification` fails in `test_acc_top5_card4`, I found that it is the original error of CE Debug, since there is [another task](http://ce.paddlepaddle.org:8080/viewLog.html?buildId=56696&buildTypeId=PaddleModesl_BuildModels&tab=buildResultsDiv) also failing.

### CODE REVIEW

File: paddle/fluid/memory/allocation/allocator_facade.cc [hunk 1]
```diff
@@ -37,7 +37,7 @@
 #endif
 
 DEFINE_int64(
-    gpu_allocator_retry_time, 0,
+    gpu_allocator_retry_time, 100,
```
Comment (chengduoZH) [inline paddle/fluid/memory/allocation/allocator_facade.cc:40]: Why do you set 100 here?

File: paddle/fluid/memory/allocation/cuda_allocator.cc [hunk 2]
```diff
@@ -36,7 +37,7 @@ Allocation* CUDAAllocator::AllocateImpl(size_t size) {
   void* ptr;
   auto status = cudaMalloc(&ptr, size);
   if (UNLIKELY(status != cudaSuccess)) {
-    throw BadAlloc(string::Sprintf(
+    PADDLE_THROW_BAD_ALLOC(string::Sprintf(
```
Comment (chengduoZH) [inline paddle/fluid/memory/allocation/cuda_allocator.cc]: Why do you use `string::Sprintf` here?

File: paddle/fluid/memory/allocation/allocator_facade.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator_facade.cc:40]: I am not sure what this value should be. TensorFlow sets it to be 10 seconds, but I think it is too large...

File: paddle/fluid/memory/allocation/cuda_allocator.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/cuda_allocator.cc]: `string::Sprintf` can be removed here. Done.

File: paddle/fluid/memory/detail/system_allocator.cc [hunk 3]
```diff
@@ -118,19 +119,19 @@ void* GPUAllocator::Alloc(size_t* index, size_t size) {
...
-                 << FLAGS_initial_gpu_memory_in_mb
-                 << ". Current FLAGS_reallocate_gpu_memory_in_mb value is "
-                 << FLAGS_reallocate_gpu_memory_in_mb;
-    return nullptr;
+    PADDLE_THROW_BAD_ALLOC(
+        "Cannot malloc " + std::to_string(size / 1024.0 / 1024.0) +
+        " MB GPU memory. Please shrink "
+        "FLAGS_fraction_of_gpu_memory_to_use or "
+        "FLAGS_initial_gpu_memory_in_mb or "
+        "FLAGS_reallocate_gpu_memory_in_mb"
```
Comment (zhhsplendid) [inline paddle/fluid/memory/detail/system_allocator.cc:126]: No space here, "in_mb" will concat with "environment" next line

File: paddle/fluid/platform/enforce.h [hunk 4]
```diff
@@ -66,25 +66,59 @@ inline std::string demangle(std::string name) {
 inline std::string demangle(std::string name) { return name; }
 #endif
 
+template <typename StrType>
+inline std::string GetTraceBackString(StrType&& what, const char* f, int l) {
```
Comment (zhhsplendid) [inline paddle/fluid/platform/enforce.h]: "f", "l" is not good parameter names. In my understanding, you print trace at [file:line] Could you change "f" to "file" or "file_name", "l" to "line" or "line_number" or something like that.

File: paddle/fluid/memory/allocation/allocator_facade.cc [hunk 5]
```diff
@@ -145,6 +153,25 @@ class AllocatorFacadePrivate {
...
+      PADDLE_ENFORCE_EQ(pair.second->IsAllocThreadSafe(), true);
+    }
+  }
+
+  void WrapCUDARetryAllocator(size_t retry_time) {
+    PADDLE_ENFORCE_GT(retry_time, 0, "Retry time must be larger than 0");
+    for (auto& pair : allocators_) {
+      if (!platform::is_gpu_place(pair.first)) continue;
+
+      pair.second = std::make_shared<RetryAllocator>(pair.second, retry_time);
```
Comment (zhhsplendid) [inline paddle/fluid/memory/allocation/allocator_facade.cc]: Make it shorter without "continue" logic:

if (platform::is_gpu_place(pair.first)) {
    pair.second = std::make_shared<RetryAllocator>(pair.second, retry_time);
}

File: paddle/fluid/memory/allocation/retry_allocator.cc [hunk 6]
```diff
@@ -31,29 +53,29 @@ Allocation* RetryAllocator::AllocateImpl(size_t size) {
   // But it would add lock even when allocation success at the first time
   try {
     return alloc_func();
-  } catch (BadAlloc& bad_alloc) {
+  } catch (BadAlloc&) {
```
Comment (zhhsplendid) [inline paddle/fluid/memory/allocation/retry_allocator.cc:60]: Consider adding some logs when catching BadAlloc. Same as while loop beneath. It may be easier to debug in the future.

**Review by chengduoZH — DISMISSED** [review]
LGTM

File: paddle/fluid/memory/detail/system_allocator.cc (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/memory/detail/system_allocator.cc:126]: Discuss offline. Maybe we should change the scripts in benchmark repo first. This PR may not change this log.

File: paddle/fluid/platform/enforce.h (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/platform/enforce.h]: Done.

File: paddle/fluid/memory/allocation/allocator_facade.cc (hunk 5 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator_facade.cc]: Done.

File: paddle/fluid/memory/allocation/retry_allocator.cc (hunk 6 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/retry_allocator.cc:60]: Done.

File: paddle/fluid/memory/allocation/allocator.h [hunk 7]
```diff
@@ -19,21 +19,21 @@
...
 
 namespace paddle {
 namespace memory {
 namespace allocation {
 
 // Exception when `Alloc`/`AllocShared` failed
-class BadAlloc : public std::exception {
- public:
-  inline explicit BadAlloc(std::string msg) : msg_(std::move(msg)) {}
+struct BadAlloc : public std::exception {
```
Comment (zhhsplendid) [inline paddle/fluid/memory/allocation/allocator.h:30]: I found that there are not only one place using f, l ... including old code ... please just modify the f -> file, l -> line at newly added code. You don't have to change old code.

Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator.h:30]: Done. I have changed the old codes too.

**Review by zhhsplendid — DISMISSED** [review]
LGTM

File: paddle/fluid/memory/allocation/retry_allocator_test.cc [hunk 8]
```diff
@@ -93,6 +97,50 @@ TEST(RetryAllocator, RetryAllocator) {
   }
 }
 
+class DummyAllocator : public Allocator {
+ public:
+  bool IsAllocThreadSafe() const override { return true; }
+
+ protected:
+  Allocation *AllocateImpl(size_t size) override {
```
Comment (zhhsplendid) [inline paddle/fluid/memory/allocation/retry_allocator_test.cc:105]: Great dependency injection for test! Nothing to change. Just applause.

File: paddle/fluid/memory/allocation/retry_allocator_test.cc [hunk 9]
```diff
@@ -93,6 +97,50 @@ TEST(RetryAllocator, RetryAllocator) {
...
+    PADDLE_THROW_BAD_ALLOC("Always BadAlloc");
+  }
+
+  void FreeImpl(Allocation *) override {}
+};
+
+TEST(RetryAllocator, RetryAllocatorLastAllocFailure) {
+  {
+    RetryAllocator allocator(std::make_shared<DummyAllocator>(), 10);
+    try {
```
Comment (zhhsplendid) [inline paddle/fluid/memory/allocation/retry_allocator_test.cc:116]: Very tiny issue. You can decide whether to change or not.

Usually, actual values of variables (such as "10" here, "true/false" in some code) can decrease readability of code. A software engineer has to look at API if he/she doesn't know the second parameter of RetryAllocator is "retry_ms".

some people use
"size_t retry_ms=10; // may also be const at head of file
RetryAllocator allocator(std::make_shared<DummyAllocator>(), retry_ms)" 
or "RetryAllocator allocator(std::make_shared<DummyAllocator>(), /* retry_ms = */ 10)"

However, it depends on programmer's background knowledge whether to add it. For example, in the next line, "allocator.Allocate(100)", since "malloc" is very common in C/C++, so we don't have to add that and SWE will have sense it's byte size.

In summary, consider adding readability of code for future developer, but not too much. In this case. I suggest adding "/* retry_ms = */ 10", but keep "allocator.Allocate(100)".

**Review by zhhsplendid — DISMISSED** [review]

File: paddle/fluid/memory/allocation/retry_allocator_test.cc (hunk 8 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/retry_allocator_test.cc:105]: Thanks!

File: paddle/fluid/memory/allocation/retry_allocator_test.cc (hunk 9 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/retry_allocator_test.cc:116]: Done.

**Review by chengduoZH — APPROVED** [review]
LGTM
